File: 递归实现指数型枚举.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    cin >> n;
    vector<int> path;

    function<void(int)> dfs = [&](int pos) {
        for (int i = 0; i < path.size(); i++) {
            if (i and path[i] < path[i - 1]) {
                return;
            }
        }
        for (int i = 0; i < path.size(); i++) {
            cout << path[i] << " ";
        }
        cout << endl;
        for (int i = pos; i <= n; i++) {
            path.push_back(i);
            dfs(i + 1);
            path.pop_back();
        }
    };

    dfs(1);
    return 0;
}

File: 递归实现指数型枚举.py
n = int(input())
path = []

def dfs(pos):
    for i, x in enumerate(path):
        if i and x < path[i - 1]:
            return
    print(*path)
    for i in range(pos, n + 1):
        path.append(i)
        dfs(i + 1)
        path.pop()

dfs(1)
